### PR TITLE
Add support for files in OPFS

### LINKED ISSUE #32: Feature: use Origin-Private Filesystem (OPFS) directory
Hey, I'm not sure how much effort this would be, so: would it be viable to support the [Origin-Private Filesystem (OPFS)](https://web.dev/origin-private-file-system/) in the `fds: []` argument to the WASI constructor? I suppose that you'd need to write a new `Directory` subclass, which wraps a [FileSystemDirectoryHandle](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemDirectoryHandle) or something like that?

```ts
// open directory in OPFS
const opfs = await navigator.storage.getDirectory();
const dir = await opfs.getDirectoryHandle("wasi-context", { create: true });

// initialize WASI shim
let wasi = new WASI(args, envs, [
  new OpenFile(new File([])), // stdin
  new OpenFile(new File([])), // stdout
  new OpenFile(new File([])), // stderr
  new WrapDirHandle(dir),     // <----- ?!
]);

...
```

As far as I can tell, supporting this API would also allow you to use `window.showDirectoryPicker()` and use an *actual* directory on the client machine for the execution context.

### PR DESCRIPTION
**Pull request by ansemjo**
This could partly fix #32 by allowing to use [`FileSystemSyncAccessHandle`s](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle) in Workers for single files opened from [OPFS](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system) as [noted in my last comment](https://github.com/bjorn3/browser_wasi_shim/issues/32#issuecomment-1612637810). Usage example:

```ts
// open synchronous handle to a file in opfs
const opfs = await navigator.storage.getDirectory();
const file = await opfs.getFileHandle("example.txt");
const handle = await file.createSyncAccessHandle();

// populate the virtual filesystem for shim
const fds: Array<Fd> = [
  new OpenFile(new File([])), // stdin
  new OpenFile(new File([])), // stdout
  new OpenFile(new File([])), // stderr
  new PreopenDirectory("/", {
    "example.txt": new OPFSFile(handle),
  }),
];
const shim = new WASI(args, envs, fds);
```

~~**This should currently be considered *very* WIP, as I've developed this out-of-tree in my project and then tried to copy it into this fork. That means I haven't actually run the code here yet.**~~ In my project I could successfully read a small file though: 
image

### CODE REVIEW

**Comment by ansemjo** [comment]
* One problem that I already noticed, is that the file is never actually closed though. My experiment in Rust just uses `std::fs::read()` to read the contents and so apparently does not explicitly call `fd_close()` anywhere. The result is that trying to access the file a second time fails with a `NoModificationAllowedError: No modification allowed`.
* Also, I should probably implement more of the `Fd` methods ... `fd_sync()` comes to mind, as the `FileSystemSyncAccessHandle` brings a `flush()` method. We should probably use that.

**Comment by bjorn3** [comment]
Thanks a lot for working on this!

> One problem that I already noticed, is that the file is never actually closed though. My experiment in Rust just uses std::fs::read() to read the contents and so apparently does not explicitly call fd_close() anywhere. The result is that trying to access the file a second time fails with a NoModificationAllowedError: No modification allowed.

Weird that it is never closed. Can you use the strace wrapper to show all syscalls that are made? In any case there should be code added to handle opening the same file multiple times. At least once directories are supported.

File: src/fs_core.ts [hunk 1]
```diff
@@ -21,19 +21,46 @@ export class File {
...
+export interface FileSystemSyncAccessHandle {
+  close(): undefined;
+  flush(): undefined;
+  getSize(): number;
+  read(buffer: ArrayBuffer | ArrayBufferView, options?: { at: number }): number;
+  truncate(to: number): undefined;
+  write(buffer: ArrayBuffer | ArrayBufferView, options?: { at: number }): number;
+}
+
+export class OPFSFile {
```
[the remark this answers is not in the document]
Comment (ansemjo) [reply src/fs_core.ts:34]: Yep, you're right! I'm currently trying to fix the exported typings, after figuring out how to `yarn link` my local fork into my project ...

**Comment by ansemjo** [comment]
> Thanks a lot for working on this!
> 
> > One problem that I already noticed, is that the file is never actually closed though. My experiment in Rust just uses std::fs::read() to read the contents and so apparently does not explicitly call fd_close() anywhere. The result is that trying to access the file a second time fails with a NoModificationAllowedError: No modification allowed.
> 
> Weird that it is never closed. Can you use the strace wrapper to show all syscalls that are made? In any case there should be code added to handle opening the same file multiple times. At least once directories are supported.

Maybe I should clarify: by opening a second time I mean starting a second WebAssembly instance, which also tries to open the file from OPFS. So technically, the previous instance simply still holds the lock on the file. My expectation would have been that at the end of `wasi.start(instance)` all opened files would be closed again but I am not sure if that'd be correct behaviour.

Anyway, how do I use the strace wrapper? Just wrap the `"wasi_snapshot_preview1"` import as seen in the example?
```
let inst = await WebAssembly.instantiate(wasm, {
  "wasi_snapshot_preview1": strace(w.wasiImport, ["fd_prestat_get"]),
});
```

**Comment by bjorn3** [comment]
> Maybe I should clarify: by opening a second time I mean starting a second WebAssembly instance, which also tries to open the file from OPFS. So technically, the previous instance simply still holds the lock on the file. My expectation would have been that at the end of wasi.start(instance) all opened files would be closed again but I am not sure if that'd be correct behaviour.

For reactor style wasi modules as handled by `.initialize()` fds should be kept. For command style wasi modules as handled by `.start()` fds could be closed at the end. That would be a breaking change though. Maybe you could add a FIXME to `start()` to do this in v0.3?

> Anyway, how do I use the strace wrapper? Just wrap the "wasi_snapshot_preview1" import as seen in the example?

Yes: https://github.com/bjorn3/browser_wasi_shim/blob/ee4f28732cd86bbebeb92969af7cee6d30ac5225/examples/rustc.html#L129-L131

**Comment by ansemjo** [comment]
Aha! Mea culpa .. the Rust experiment actually *does* close the file as seen here:

image

This is the code that is executed there:

```rust
fn main() {
  println!("This is Rust.");

  // print commandline arguments
  let args: Vec<String> = std::env::args().collect();
  let name = args[0].clone();
  println!("file '{name}' was called with arguments: {args:?}");

  // print environment variables
  println!("environment variables:");
  for (key, value) in std::env::vars() {
    println!(" - {}: {}", key, value);
  };

  // list root filesystem
  match std::fs::read_dir("/") {
    Err(e) => println!("ERR: couldn't open directory '/': {e}"),
    Ok(iter) => {
      println!("listing '/' contents:");
      for item in iter {
        println!(" - {}", item.unwrap().path().display());
      }
    }
  };

  // try to read a specific file
  const FILENAME: &str = "/hello.txt";
  match std::fs::read(FILENAME) {
    Err(e) => println!("ERR: couldn't open file '{FILENAME}': {e}"),
    Ok(content) => {
      let text = String::from_utf8_lossy(&content);
      println!("'{FILENAME}': {text}");
    },
  }

}
```

My TinyGo experiment however, does not ..

image

```go
package main

import (
	"fmt"
	"io"
	"os"
)

func main() {
	fmt.Println("This is TinyGo.")

	// print commandline arguments
	args := os.Args
	name := args[0]
	fmt.Printf("file '%s' was called with arguments: %v\n", name, args)

	// print environment variables
	fmt.Println("environment variables:")
	for _, env := range os.Environ() {
		fmt.Printf(" - %s\n", env)
	}

	// list root filesystem
	contents, err := os.ReadDir("/")
	if err != nil {
		fmt.Printf("ERR: couldn't open directory '/': %s\n", err)
	} else {
		fmt.Println("listing '/' contents:")
		for _, item := range contents {
			fmt.Printf(" - %s\n", item.Name())
		}
	}

	// try to read a specific file
	const FILENAME = "/hello.txt"
	file, err := os.Open(FILENAME)
	if err != nil {
		fmt.Printf("ERR: couldn't open file '%s': %s\n", FILENAME, err)
	} else {
		bytes, err := io.ReadAll(file)
		if err != nil {
			fmt.Printf("ERR: couldn't read file '%s': %s\n", FILENAME, err)
		} else {
			fmt.Printf("'%s': %s\n", FILENAME, string(bytes))
		}
	}

}
```

I mean, yeah .. I neither call nor defer `file.Close()` anywhere. I've added a FIXME in the last commit.

File: src/fs_fd.ts [hunk 2]
```diff
@@ -178,6 +262,8 @@ export class OpenDirectory extends Fd {
     if (entry instanceof File) {
       // @ts-ignore
       return { ret: 0, fd_obj: new OpenFile(entry) };
+    } else if (entry instanceof OPFSFile) {
+      return { ret: 0, fd_obj: new OpenOPFSFile(entry) }
```
[the remark this answers is not in the document]
Comment (ansemjo) [reply src/fs_fd.ts]: Don't they all already derive from `Fd`? We could just use duck-typing to check if the `entry` has an `open()` property and call that.

[the remark this answers is not in the document]
Comment (ansemjo) [reply src/fs_fd.ts]: Well, none of `File | SyncOPFSFile | Directory` derived from `Fd` but Typescript handles the type union to check if `.open()` exists just fine without a common base class: 512a46d

**Comment by ansemjo** [comment]
It is definitely incorrect to close the handle in `fd_close()` though, as that makes the file unusable for the rest of the execution. So this should also be handled in `SyncOPFSFile` after `WASI.start()`.

File: src/fs_core.ts [hunk 3]
```diff
@@ -37,10 +39,13 @@ export interface FileSystemSyncAccessHandle {
 }
 
 export class SyncOPFSFile {
+  // FIXME needs a close() method to be called after start() to release the underlying handle
   constructor(public handle: FileSystemSyncAccessHandle) { }
 
-  open() {
-    return new OpenSyncOPFSFile(this);
+  open(fd_flags: number) {
```
Comment (ansemjo) [inline src/fs_core.ts:45]: @bjorn3 Added handling for `FDFLAGS_APPEND` here and in `File` to be able to correctly *append* to existing files.

**Comment by ansemjo** [comment]
I've tested writes and caught an OOB bug in the process; added support for `FDFLAGS_APPEND`; and removed my `console.log` lines.

Regarding the closing of the `FileSystemSyncAccessHandle`: I'm thinking that maybe we just shouldn't bother with it, since the way I have implemented it here, I already accept a handle. Which means it was created externally and maybe we *shouldn't* make assumptions and close it.

The `Inode` base class idea I'd leave for another pull request. 😛 

If you don't have any further comments, I think you might merge this now as the basic read / write / seek / append operations work now, as far as I can tell.

File: src/fs_core.ts [hunk 4]
```diff
@@ -21,19 +28,57 @@ export class File {
   }
 }
 
+// shim https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle
```
Comment (bjorn3) [inline src/fs_core.ts:31]:
```suggestion
// Shim for https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle
// This is not part of the public interface.
```

Also looks like updating to typescript 5 is enough to make this available: https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1471 That should probably be a separate PR.

File: src/index.ts [hunk 5]
```diff
@@ -2,6 +2,6 @@ import WASI from "./wasi.js";
 export { WASI };
 
 export { Fd } from './fd.js';
-export { File, Directory } from "./fs_core.js";
-export { OpenFile, OpenDirectory, PreopenDirectory } from "./fs_fd.js";
+export { File, SyncOPFSFile as OPFSFile, Directory } from "./fs_core.js";
```
Comment (bjorn3) [inline src/index.ts:5]:
```suggestion
export { File, SyncOPFSFile, Directory } from "./fs_core.js";
```

**Comment by bjorn3** [comment]
I'm going to test this myself tomorrow. If I don't see anything weird I will merge it.

File: src/fs_core.ts (hunk 4 above)
Comment (ansemjo) [reply src/fs_core.ts]: Huh. Maybe there's something else you need to set because this is only available in WebWorkers? My project is on TypeScript 5.0.4 and doesn't seem to include this type. I've already used [`@types/web`](https://www.npmjs.com/package/@types/web) to get support for the WebTransport API but the editor still complains about `FileSystemSyncAccessHandle`.

Comment (ansemjo) [reply src/fs_core.ts]: Indeed, [you need to tell Typescript](https://joshuatz.com/posts/2021/strongly-typed-service-workers/) the the file should use the `webworker` lib:

```ts
/// <reference no-default-lib="true"/>
/// <reference lib="es2020" />
/// <reference lib="webworker" />

...
```

And as soon as I did that it started complaining that the types are incompatible 🙃 
image

But then adding `/// <reference lib="webworker" />` at the top of `fd_core.ts` and installing `typescript@^5.0.4` fixes that for me. I'm not adding that as a commit though because you said that'd be for a separate PR.

I've just fixed the "shim" for now because while MDN says these return a `Promise<undefined>`, which is obviously an error, I thought they meant just `undefined`. The specification and typings seem to use `void` though.

File: src/fs_core.ts [hunk 6]
```diff
@@ -21,19 +28,58 @@ export class File {
...
+export interface FileSystemSyncAccessHandle {
+  close(): void;
+  flush(): void;
+  getSize(): number;
+  read(buffer: ArrayBuffer | ArrayBufferView, options?: { at: number }): number;
+  truncate(to: number): void;
+  write(buffer: ArrayBuffer | ArrayBufferView, options?: { at: number }): number;
+}
+
+export class SyncOPFSFile {
```
Comment (bjorn3) [inline src/fs_core.ts:42]:
```suggestion
// Synchronous access to an individual file in the origin private file system.
// Only allowed inside a WebWorker.
export class SyncOPFSFile {
```

**Comment by bjorn3** [comment]
Do you have more changes you want or shall I publish a new release with this?

**Comment by ansemjo** [comment]
> Got it working locally.

Great! 😊 

> Do you have more changes you want or shall I publish a new release with this?

Right now I'm just glad you found this useful enough to merge so quickly. :) I see you reopened the issue to track work on the async API / whole directories .. that'd be *great* but I don't think I can justify the effort to even attempt that right now.

It probably should've been part of this PR but maybe add another usage example on how to use this new `SyncOPFSFile`, since the only way to find it right now is to look through the code or find this issue. Is your test code suitable for this?

Other than that I don't have anything that should be a blocker for a new release, I think.

**Comment by bjorn3** [comment]
> I see you reopened the issue to track work on the async API / whole directories .. that'd be great but I don't think I can justify the effort to even attempt that right now.

No problem.

> It probably should've been part of this PR but maybe add another usage example on how to use this new SyncOPFSFile, since the only way to find it right now is to look through the code or find this issue. Is your test code suitable for this?

The code I used to test it was using the OPFS file as stderr for rustc.wasm and then manually verifying that it wrote to the file in the browser console. Not really something usable as test for others.
